src/operator/tensor/matrix_op-inl.h

```diff
@@ -927,6 +927,140 @@ void SliceAxis(const nnvm::NodeAttrs& attrs,
+  CHECK_NE(req, kWriteInplace) << "kWriteInplace for SliceAxis on CSR input is not supported";
+  int axis, begin, end;
+  GetSliceAxisParams(param, in.shape(), &axis, &begin, &end);
+  int indptr_len = in.shape()[0] + 1;
+  out.CheckAndAllocAuxData(kIndPtr, Shape1(indptr_len));
+  // assume idx indptr share the same type
+  MSHADOW_IDX_TYPE_SWITCH(in.aux_type(kIndPtr), RType, {
+    MSHADOW_IDX_TYPE_SWITCH(in.aux_type(kIdx), IType, {
+      MSHADOW_TYPE_SWITCH(in.dtype(), DType, {
+        RType *in_indptr = in.aux_data(kIndPtr).dptr<RType>();
+        IType *in_idx = in.aux_data(kIdx).dptr<IType>();
+        DType *in_data = in.data().dptr<DType>();
+        // retrieve nnz (CPU implementation)
+        RType *out_indptr = out.aux_data(kIndPtr).dptr<RType>();
+        int nnz = 0;
+        out_indptr[0] = 0;
+        for (int i=0; i < indptr_len - 1; i++) {
+          out_indptr[i+1] = out_indptr[i];
+          for (int j=in_indptr[i]; j < in_indptr[i+1]; j++) {
+            if (in_idx[j] >= begin && in_idx[j] < end) {
+              out_indptr[i+1]++;
+              nnz++;
+            }
+          }
+        }
+        out.CheckAndAllocAuxData(kIdx, Shape1(nnz));
+        out.CheckAndAllocData(Shape1(nnz));
+        IType *out_idx = out.aux_data(kIdx).dptr<IType>();
+        DType *out_data = out.data().dptr<DType>();
+
+        Stream<xpu> *s = ctx.get_stream<xpu>();
+        Kernel<SliceAxisOneCsrAssign, xpu>::Launch(s, indptr_len-1, out_idx, out_data,
+                                                   out_indptr, in_idx, in_data, in_indptr,
+                                                   begin, end);
+      });
+    });
+  });
+}
+
+template<typename xpu>
+void SliceAxisZeroCsrImpl(const SliceAxisParam &param, const OpContext& ctx,
+                  const NDArray &in, OpReqType req, const NDArray &out) {
+  int axis, begin, end;
+  GetSliceAxisParams(param, in.shape(), &axis, &begin, &end);
+  SliceParam slice_param;
+  slice_param.begin[0] = begin;
+  slice_param.end[0] = end;
+  SliceCsrImpl<xpu>(slice_param, ctx, in, req, out);
+}
+
+template<typename xpu>
+void SliceAxisEx(const nnvm::NodeAttrs& attrs,
+          const OpContext& ctx,
+          const std::vector<NDArray>& inputs,
+          const std::vector<OpReqType>& req,
+          const std::vector<NDArray>& outputs) {
+  CHECK_EQ(inputs.size(), 1);
+  CHECK_EQ(outputs.size(), 1);
+
+  const SliceAxisParam& param = nnvm::get<SliceAxisParam>(attrs.parsed);
+  auto in_stype = inputs[0].storage_type();
+  CHECK_NE(in_stype, kDefaultStorage)
+           << "SliceAxisEx is not expected to execute for input with default storage type";
+  if (in_stype == kCSRStorage) {
+    if (param.axis == 0) {
+      SliceAxisZeroCsrImpl<xpu>(param, ctx, inputs[0], req[0], outputs[0]);
+    } else if (param.axis == 1) {
+      SliceAxisOneCsrImpl<xpu>(param, ctx, inputs[0], req[0], outputs[0]);
+    } else {
+      LOG(FATAL) << "CSRNDArray is only for 2-D shape";
```

Comment:
Does it fail with negative axis? I think GetSliceAxisParams already handles negative axis for you